From 2725e86afa026ffdc0a811e726e19fe347863340 Mon Sep 17 00:00:00 2001
From: chandavath <rcrajagopal@habana.ai>
Date: Mon, 18 Mar 2024 07:26:15 +0200
Subject: [PATCH] [SW-137121] gaudi3: erom pcie: added cold flops mechanism

In erom, we bringup PCIe link up through ARC1, not through
PIF ARC. To indicate this to ppboot, added cold flops mechanism.

So that ppboot, can read the cold flops and warn the users
PIF ARC FW and PCIe PHY FW are not running, to bring it
up, perform cold boot.

Change-Id: I21af54c11ac021fe3ce2e421ecbabec802a9015e
Signed-off-by: chandavath <rcrajagopal@habana.ai>
---

diff --git a/app/ppboot/ppboot.c b/app/ppboot/ppboot.c
index bd4f9be..cdc3f5d 100644
--- a/app/ppboot/ppboot.c
+++ b/app/ppboot/ppboot.c
@@ -317,6 +317,7 @@
 __weak void hl_pcie_enable_elbi_access(void) {}
 __weak void ppboot_tpm_extend(void) {}
 __weak void ppboot_boot_info_init(void) {}
+__weak void hl_check_pcie_link_up_by_erom(void) {}
 
 void ppboot_main(void)
 {
@@ -359,6 +360,7 @@
 	ppboot_time_critical_section_end();
 	/**********time critical section end************/
 	ppboot_pcie_drain_init();
+	hl_check_pcie_link_up_by_erom();
 	ppboot_version();
 	rst_src_dump();
 
diff --git a/arch/gaudi3/arc/common.c b/arch/gaudi3/arc/common.c
index 6ede63c..c9c4019 100644
--- a/arch/gaudi3/arc/common.c
+++ b/arch/gaudi3/arc/common.c
@@ -1554,3 +1554,16 @@
 
 	return 0;
 }
+
+void hl_check_pcie_link_up_by_erom(void)
+{
+	struct gaudi3_cold_rst_data cold_rst_data;
+
+	cold_rst_data.data = regmap_readl(
+		D0_PSOC_GLOBAL_CONF_REG(mmGLOBAL_CONF_COLD_RST_FLOPS_2));
+
+		if (cold_rst_data.pcie_bringup_by_erom) {
+			preboot_puts_raw("Warning: PCIe link may not be at Gen5 speed, ");
+			preboot_puts("please perform cold reset\n");
+		}
+}
diff --git a/arch/gaudi3/common/pci.c b/arch/gaudi3/common/pci.c
index 39d4df4..97f186b 100644
--- a/arch/gaudi3/common/pci.c
+++ b/arch/gaudi3/common/pci.c
@@ -306,6 +306,19 @@
 	}
 }
 
+static void hl_set_pcie_bringup_by_erom(void)
+{
+	struct gaudi3_cold_rst_data cold_rst_data;
+
+	cold_rst_data.data = regmap_readl(
+		D0_PSOC_GLOBAL_CONF_REG(mmGLOBAL_CONF_COLD_RST_FLOPS_2));
+
+	cold_rst_data.pcie_bringup_by_erom = 0x1;
+
+	regmap_writel(cold_rst_data.data,
+		D0_PSOC_GLOBAL_CONF_REG(mmGLOBAL_CONF_COLD_RST_FLOPS_2));
+}
+
 void hl_erom_bringup_pcie_link(void)
 {
 	u8 ltssm_en;
@@ -358,6 +371,8 @@
 		hl_pcie_clear_boot_termination();
 
 		regmap_writel(0x3, D0_PCIE_AUX_REG(mmPCIE_AUX_LTSSM_EN));
+
+		hl_set_pcie_bringup_by_erom();
 	} else {
 		hl_pcie_clear_boot_termination();
 	}
diff --git a/include/baremetal.h b/include/baremetal.h
index 1821b17..4dcdef6 100644
--- a/include/baremetal.h
+++ b/include/baremetal.h
@@ -289,6 +289,7 @@
 void hl_erom_bringup_pcie_link(void);
 void hl_erom_boot_info_set_version(u32 type, void *data, u32 size);
 int hl_boot_info_finalize(void);
+void hl_check_pcie_link_up_by_erom(void);
 
 #endif /* __ASSEMBLY__ */
 
